# Importing Libraries

In [5]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Import Data

In [6]:
# Folder path

path = r'C:\Users\balaj\Documents\Data Analytics Course\Instacart Basket Analysis'

In [7]:
# Import products.csv

df_prod = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [11]:
# Import orders_wrangled.csv 

df_ord = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [10]:
df_ord.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

# Step 3: Consistency Check

Mixed Data Types

In [12]:
# Create a Dataframe
df_test = pd.DataFrame()

In [13]:
# Create a mixed column
df_test['mix'] = ['a','b', 1, True]

In [14]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [15]:
# Changing 'mix' to string type:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [16]:
# Count the missing vailes in each col:
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# Missing Values list: 
df_nanprod = df_prod[df_prod.product_name.isnull() == True]

In [18]:
# Create df without nan values:
df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]

In [19]:
df_prod_clean.shape

(49677, 5)

# Duplicates

In [20]:
# Create duplicates df:
df_dups = df_prod_clean[df_prod_clean.duplicated()]

In [21]:
# Create dropped dups list 
df_prod_clean_no_dups = df_prod_clean.drop_duplicates()

In [22]:
df_prod_clean_no_dups.shape

(49672, 5)

# Export changes

In [24]:
# Export changes df:
df_prod_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Data Constistency Check

In [25]:
# Step 2. Run the df.describe() function on your df_ords dataframe:
df_prod.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Question 2: Min value = 1.000000, Max value 99999.000000, perhaps this could been a error.

In [26]:
df_prod[df_prod.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [27]:
# Step 3. Check for mixed-type data in your df_ords dataframe:
for col in df_ord.columns.tolist():
  weird = (df_ord[[col]].applymap(type) != df_ord[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
order_day_of_week  consistent
order_hour_of_day  consistent
days_since_last_order  consistent


In [28]:
# Step 4. If you find mixed-type data, fix it:

The column in question should contain observations of a single data type: No mixed-type data found.

In [30]:
# Step 5. Run a check for missing values in your df_ords dataframe:
df_ord.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [31]:
df_ord.head(20)

Unnamed: 0  order_id  user_id eval_set  order_number  order_day_of_week  \
0            0   2539329        1    prior             1                  2   
1            1   2398795        1    prior             2                  3   
2            2    473747        1    prior             3                  3   
3            3   2254736        1    prior             4                  4   
4            4    431534        1    prior             5                  4   
5            5   3367565        1    prior             6                  2   
6            6    550135        1    prior             7                  1   
7            7   3108588        1    prior             8                  1   
8            8   2295261        1    prior             9                  1   
9            9   2550362        1    prior            10                  4   
10          10   1187899        1    train            11                  4   
11          11   2168274        2    prior             1                  2   
12          12   1501582        2    prior             2                  5   
13          13   1901567        2    prior             3                  1   
14          14    738281        2    prior             4                  2   
15          15   1673511        2    prior             5                  3   
16          16   1199898        2    prior             6                  2   
17          17   3194192        2    prior             7                  2   
18          18    788338        2    prior             8                  1   
19          19   1718559        2    prior             9                  2   

    order_hour_of_day  days_since_last_order  
0                   8                    NaN  
1                   7                   15.0  
2                  12                   21.0  
3                   7                   29.0  
4                  15                   28.0  
5                   7                   19.0  
6                   9                   20.0  
7                  14                   14.0  
8                  16                    0.0  
9                   8                   30.0  
10                  8                   14.0  
11                 11                    NaN  
12                 10                   10.0  
13                 10                    3.0  
14                 10                    8.0  
15                 11                    8.0  
16                  9                   13.0  
17                 12                   14.0  
18                 15                   27.0  
19                  9                    8.0

In [33]:
df_ord['user_id'] = df_ord['user_id'].astype('int64')
df_ord.user_id.max()

206209

In [36]:
# create a dataframe for the missing values
df_ord_nan = df_ord [df_ord ['days_since_last_order'].isnull () == True]

In [37]:
df_ord_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         order_day_of_week  order_hour_of_day  days_since_last_order  
0                        2                  8                    NaN  
11                       2                 11                    NaN  
26                       1                 14                    NaN  
39                       6                 11                    NaN  
45                       3                 12                    NaN  
...                    ...                ...                    ...  
3420930                  4                 12                    NaN  
3420934                  3                 18                    NaN  
3421002                  6                 19                    NaN  
3421019                  1                 15                    NaN  
3421069                  3                 11                    NaN  

[206209 rows x 8 columns]

The reason for the missing value is perhaps because the majority of customers are first-time clients who have never placed an order before. In addition, there won't be a value indicating how long it has been since a prior order. The maximum value in the column 'user_id' is 206209, which corresponds with the number of missing values in the column 'days_since_prior_order" that represents one for each new customer.

In [38]:
# Step 6. Address the missing values using an appropriate method:

I’ve chosen to make no changes. This is because I believe that it isn’t reasonable to assign values or remove or filter the data. Statistics may still be analyzed with the NaN values present. A value of "1’ in the column 'order_number' can represent new customers.

In [40]:
# Step 7. Run a check for duplicate values in your df_ords data: 
df_ord_dups = df_ord[df_ord.duplicated()]

In [41]:
df_ord_dups.size

0

In [43]:
# No duplicates found

In [44]:
# Step 8. Address the duplicates using an appropriate method:

I’ve chosen to make no changes, following step 6 action.

In [45]:
# Step 9. Export changes: 
df_ord.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))